joomla/cassiopeia
# [Metismenu] Fix dropdown toggles and other fixes for metismenu

Pull Request for Issue #125 .

Replaces pull request (PR) #184 .

### Summary of Changes

This pull request (PR) changes following on the metismenu for menu items which are parent items, i.e. have sub-menu items below them, i.e. have a drop-down submenu:

- Move the creation of the dropdown toggle button from the metismenu.php layout to the sub-layouts.
- In case of menu items **_without_** a hyperlink, i.e. items of type "heading" or "separator", let the complete menu item be the dropdown toggle button. This solves the main topic of issue #125 , clickable area on dropdown menu.
- Adapt (s)css to the previous change.
- In case of menu items **_with_** a hyperlink, i.e. items of type "component" or "url", use `aria-label="' . $item->title . '"` with `$item->title` being the menu item's title instead of `<span class="sr-only">' . Text::_('JGLOBAL_TOGGLE_DROPDOWN') . '</span>`. This solved screen readers reading always the same text for all closed dropdowns, see [https://github.com/joomla/cassiopeia/issues/125#issuecomment-698602906](https://github.com/joomla/cassiopeia/issues/125#issuecomment-698602906).
- Correct `aria-haspopup` and `aria-expanded` attributes for menu items **_with_** a hyperlink. Right now `aria-haspopup` is added to the hyperlink and not to the dropdown toggle, and `aria-expanded` is added to both the hyperlink and the dropdown toggle. This PR adds both `aria-haspopup` and `aria-expanded` only to the dropdown toggle but not to the hyperlink.
- Menu items **_without_** a hyperlink are underlined completely when hovering or active.
- The title of menu items **_with_** a hyperlink is underlined separately from the dropdown toggle when hovering.

In addition, this PR changes following on the metismenu for menu items with or without a dropdown:
- Menu items **_without_** a hyperlink have an empty `title` attribute. This PR here corrects it by not adding this attribute if it is empty in the same way as it is already done with menu items **_with_** a hyperlink.

Finally this PR does an unrelated code-smell change: It removes the unnecessary `break;` at the default case of as `switch` statement where the default case is the last case in that `switch`.

### Testing Instructions

If you have development environment for Joomla 4, checkout the [development-metismenu-dropdown-fixes-1](https://github.com/joomla/cassiopeia/tree/development-metismenu-dropdown-fixes-1) branch, pull the latest changes, run "composer install" and "npm ci", and then make a new installation.

If you don't have that, use the following package to make a new installation: (obsolete link removed)

After that, login to backend and install "Blog Sample Data" to see the metismenu being used in frontend.

Then play around with different types of menu items with and without link and with and without a pulldown submenu on different menu levels.

Add more menu items to have a deeper menu.

Move the module to the sidebar and see if it works well and looks good there, too.

### Actual result BEFORE applying this Pull Request

#### Menu items without a hyperlink, i.e. of type heading or separator, and with a dropdown:

Behavior:
1. When hovering the menu item, the complete menu item (i.e. title and dropdown toggle) gets underlined. This is wrong because only the dropdown toggle is "clickable" but not the menu item title, see issue #125 .
2. The mouse pointer changes to a text cursor when hovering the menu item title. This at least confusing.
3. The mouse pointer changes to "hand" pointer when hovering the dropdown toggle like it is with hyperlinks. This is correct for a button as well, so that's ok.
4. The menu item's title (span element) has an empty title attribute.

Markup:
```
<li class="metismenu-item item-105 deeper parent">
	<span title="" class="mod-menu__heading nav-header" aria-haspopup="true" aria-expanded="false">Sample Layouts</span>
	<button class="mm-collapsed mm-toggler" aria-expanded="false"><span class="sr-only">Toggle Dropdown</span></button>
	<ul class="mm-collapse">
	...
	</ul>
</li>
```

#### Menu items with a hyperlink, i.e. of type component or url, and with a dropdown:

Behavior:
1. When hovering the menu item title or dropdown toggle, the complete menu item (i.e. title and dropdown toggle together) gets underlined. This is wrong because they are different elements with different actions when being selected.
2. When the dropdown is closed, screen readers will read "toggle dropdown - button - collapsed" for every closed dropdown, which unacceptable for a11y especially when having more than one dropdown on the same level. See [https://github.com/joomla/cassiopeia/issues/125#issuecomment-698602906](https://github.com/joomla/cassiopeia/issues/125#issuecomment-698602906).

Markup:
```
<li class="metismenu-item item-105 deeper parent">
	<a href="/index.php/sample-layouts" aria-haspopup="true" aria-expanded="false">Sample Layouts</a>
	<button class="mm-collapsed mm-toggler" aria-expanded="false"><span class="sr-only">Toggle Dropdown</span></button>
	<ul class="mm-collapse">
	...
	</ul>
</li>
```

### Expected result AFTER applying this Pull Request

#### Menu items without a hyperlink, i.e. of type heading or separator, and with a dropdown:

Behavior:
1. When hovering the menu item, the complete menu item (i.e. title and dropdown toggle "icon") gets underlined. This is now correct because the complete menu item "clickable".
2. The mouse pointer changes to a "hand" pointer when hovering the menu item.
3. When one of the submenu items is active, the menu item is underlined, too, like it is with the default layout.
4. The menu item is a button element now. This has no empty title.

Markup:
```
<li class="metismenu-item item-105 deeper parent">
	<button class="mod-menu__heading nav-header mm-collapsed mm-toggler" aria-haspopup="true" aria-expanded="false">Sample Layouts</button>
	<ul class="mm-collapse">
	...
	</ul>
</li>
```

#### Menu items with a hyperlink, i.e. of type component or url, and with a dropdown:

Behavior:
1. Menu item title and dropdown toggle are underlined independently from each other when one of them is being hovered.
2. The mouse pointer changes to a "hand" pointer when hovering the menu item title or the dropdown toggle.
3. When one of the submenu items is active, menu item title and dropdown toggle are both underlined separately.
4. For screen readers there is an aria-label for the dropdown and not the always  same "Toggle Dropdown" anymore.

Markup:
```
<li class="metismenu-item item-105 deeper parent">
	<a href="/index.php/sample-layouts">Sample Layouts</a>
	<button class="mm-collapsed mm-toggler" aria-haspopup="true" aria-expanded="false" aria-label="Sample Layouts"></button>
	<ul class="mm-collapse">
	...
	</ul>
</li>
```

### Documentation Changes Required

None.
